### PR TITLE
New tag/mech pair `foliage_color`

### PR DESCRIPTION
**Pull request by BreadcrumbIsTaken**
## `foliage_color` mechanism and tag

Adds new mechanism to change foliage (some tree leaves and vines) colors.
Adds new tag to get the foliage colors. Some biomes don't have a change in color so they just return black.

The mechanism needs to be reapplied every time the server starts and when the mechanism is applied the clients will have to be reset (leaving and joining the server)

Tested on 1.17.1, 1.18.2, and 1.19.4

This is my first time working with NMS so let me know what mistakes I've made or what can be improved!

This was fun to try and figure out :>

Requested by Udnek on [Discord](https://discord.com/channels/315163488085475337/1094959580582252674/1094959580582252674)

### CODE REVIEW

File: v1_19/src/main/java/com/denizenscript/denizen/nms/v1_19/impl/BiomeNMSImpl.java [hunk 1]
```diff
@@ -130,6 +135,24 @@ public void setPrecipitation(DownfallType type) { // TODO: 1.19.4: This is no lo
         setClimate(nmsType, getTemperature(), getTemperatureModifier(), getHumidity());*/
     }
 
+    @Override
+    public void setFoliageColor(int color) {
+        try {
+            BiomeSpecialEffects.Builder builder = (BiomeSpecialEffects.Builder) BIOMESPECIALEFFECTS_BUILDER_CONSTRUCTOR.invoke();
```
Comment (mcmonkey4eva) [inline v1_19/src/main/java/com/denizenscript/denizen/nms/v1_19/impl/BiomeNMSImpl.java]: Why is this being invoked by reflection?

File: plugin/src/main/java/com/denizenscript/denizen/objects/BiomeTag.java [hunk 2]
```diff
@@ -282,6 +283,41 @@ else if (attribute.startsWith("water", 2)) {
...
+        // # Now the leaves and vines will be a nice salmon-pink!
+        // on server start:
+        // - adjust <biome[plains]> foliage_color:<&color[#F48D8D]>
+        // -->
+        tagProcessor.registerMechanism("foliage_color", false, (object, mechanism) -> {
+            if (!mechanism.requireObject(ColorTag.class)) {
+                return;
+            }
+            object.biome.setFoliageColor(mechanism.valueAsType(ColorTag.class).asRGB());
+        });
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/BiomeTag.java]: Should use the register-with-param method

File: plugin/src/main/java/com/denizenscript/denizen/objects/BiomeTag.java [hunk 3]
```diff
@@ -282,6 +283,41 @@ else if (attribute.startsWith("water", 2)) {
...
+        // Sets the foliage color of this biome. Foliage includes leaves and vines.
+        // Can be changed once until next server restart. Colors reset on server restart.
+        // For the change to take effect on the players' clients, they must quit and rejoin the server.
+        // @tags
+        // <BiomeTag.foliage_color>
+        // @example
+        // # Adjusts the foliage color of the plains biome permanently, using a server start event to keep it applied.
+        // # Now the leaves and vines will be a nice salmon-pink!
+        // on server start:
+        // - adjust <biome[plains]> foliage_color:<&color[#F48D8D]>
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/BiomeTag.java]: Arrrrre you sure that's correct? `&color` is for chat colors, not `ColorTag` - technically you'd use `ColorTag`'s constructor but in this case can probably input the hex code directly as it's a valid `ColorTag`.

File: v1_19/src/main/java/com/denizenscript/denizen/nms/v1_19/impl/BiomeNMSImpl.java [hunk 4]
```diff
@@ -130,6 +135,24 @@ public void setPrecipitation(DownfallType type) { // TODO: 1.19.4: This is no lo
...
+    public void setFoliageColor(int color) {
+        try {
+            BiomeSpecialEffects.Builder builder = (BiomeSpecialEffects.Builder) BIOMESPECIALEFFECTS_BUILDER_CONSTRUCTOR.invoke();
+            // fogColor, waterColor, waterFogColor, and skyColor are needed for the Builder class.
+            builder.fogColor(biomeHolder.value().getFogColor())
+                    .waterColor(biomeHolder.value().getWaterColor())
+                    .waterFogColor(biomeHolder.value().getWaterFogColor())
+                    .skyColor(biomeHolder.value().getSkyColor())
+                    .foliageColorOverride(color);
+            BiomeSpecialEffects effects = builder.build();
```
Comment (tal5) [inline v1_19/src/main/java/com/denizenscript/denizen/nms/v1_19/impl/BiomeNMSImpl.java]: Won't that loose other values stored on `BiomeSpecialEffects` (I.e. background sounds, particles, etc.)?
Can just copy them all over, although maybe consider just setting the `foliageColorOverride` field on the biome's existing `BiomeSpecialEffects` object - there is a public getter for it and it'd save having to copy everything over.

**Comment by BreadcrumbIsTaken** [comment]
Pushed new changes :D

Note: the foliage color is not 100% accurate with a slight color difference. It is not very noticeable, but the difference is still there. 

Generated from gradient code:
Foliage Color Generated


Minecraft foliage map (edited with only the bottom left shown):
Foliage Color Minecraft

File: plugin/src/main/java/com/denizenscript/denizen/objects/BiomeTag.java [hunk 5]
```diff
@@ -282,6 +283,39 @@ else if (attribute.startsWith("water", 2)) {
...
+            return new ColorTag(ColorTag.fromRGB(object.biome.getFoliageColor()));
+        });
+
+        // <--[mechanism]
+        // @object BiomeTag
+        // @name foliage_color
+        // @input ColorTag
+        // @description
+        // Sets the foliage color of this biome. Foliage includes leaves and vines.
+        // Can be changed once until next server restart. Colors reset on server restart.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/BiomeTag.java]: Why only once?

File: v1_17/src/main/java/com/denizenscript/denizen/nms/v1_17/impl/BiomeNMSImpl.java [hunk 6]
```diff
@@ -78,6 +84,26 @@ public List<EntityType> getWaterEntities() {
         return getSpawnableEntities(MobCategory.WATER_CREATURE);
     }
 
+    @Override
+    public int getFoliageColor() {
+        // Check if the biome already has a default color
+        if (biomeBase.getFoliageColor() != 0) {
+            return biomeBase.getFoliageColor();
+        }
+
```
Comment (mcmonkey4eva) [inline v1_17/src/main/java/com/denizenscript/denizen/nms/v1_17/impl/BiomeNMSImpl.java]: stray newlines

**Comment by BreadcrumbIsTaken** [comment]
Done

**Pull request merged**